utils/cryptography.go

```diff
@@ -0,0 +1,325 @@
+	if err != nil {
+		return nil, fmt.Errorf("Unable to interpret jwk.Y coordinate as byte slice: %w", err)
+	}
+	pubKey.Y = new(big.Int).SetBytes(bsY)
+
+	// Step 2 decide if key is to be private
+
+	var keyUsage KeyUse
+	var privateFlag bool
+	if slices.Contains(jwk.Key_ops, "deriveKey") {
+		keyUsage = ECDH
+	} else {
+		keyUsage = ECDSA
+	}
+	if len(jwk.D) != 0 {
+		privateFlag = true
+	}
+
+	privKey := new(ecdsa.PrivateKey)
+	if privateFlag {
+		privKey.PublicKey = *pubKey
+		bsD, err := base64.URLEncoding.DecodeString(jwk.D)
+		if err != nil {
+			return nil, fmt.Errorf("Unable to interpret jwk.D coordinate as byte slice: %w", err)
+		}
+		privKey.D = new(big.Int).SetBytes(bsD)
+	}
+
+	if keyUsage == ECDSA && !privateFlag { // return an *ecdsa.PublicKey
+		return pubKey, nil
+	} else if keyUsage == ECDSA && privateFlag { // return an *ecdsa.PrivateKey
+		return privKey, nil
+	} else if keyUsage == ECDH && !privateFlag { // return an *ecdh.PublicKey
+		publicKey, err := pubKey.ECDH()
+		if err != nil {
+			return nil, err
+		}
+		return publicKey, nil
+	} else if keyUsage == ECDH && privateFlag { // return an *ecdsa.PrivateKey
+		privateKey, err := privKey.ECDH()
+		if err != nil {
+			return nil, err
+		}
+		return privateKey, nil
+	}
+	return nil, fmt.Errorf("Unable to Export key. Unrecognized Key_ops.")
+}
+
+func (privateKey *JWK) GetECDHSharedSecret(publicKey *JWK) (*JWK, error) {
+	// is public key public?
+	if len(publicKey.D) != 0 {
+		return nil, fmt.Errorf("Function takes a public JWK as argument. Private key detected.")
+	}
+	// is publis key for ECDH?
+	if !slices.Contains(publicKey.Key_ops, "deriveKey") {
+		return nil, fmt.Errorf("The public JWK passed in as argument must have 'deriveKey' as one of the key_ops")
+	}
+	// is private private?
+	if len(privateKey.D) == 0 {
+		return nil, fmt.Errorf("Function receiver must be private JWK. No private key detected.")
+	}
+	// is private for ECDH?
+	if !slices.Contains(privateKey.Key_ops, "deriveKey") {
+		return nil, fmt.Errorf("Function receiver must have 'deriveKey' as one of the Key_ops.")
+	}
+	// convert both to *ecdh.[private|public]Key
+	privKey_unCasted, err := privateKey.ExportKeyAsGoType()
+	if err != nil {
+		fmt.Println(err.Error())
+	}
+
+	var privKey *ecdh.PrivateKey
+	if pk, ok := privKey_unCasted.(*ecdh.PrivateKey); ok {
+		privKey = pk
+	}
+
+	pubKey_unCasted, err := publicKey.ExportKeyAsGoType()
+	if err != nil {
+		fmt.Println(err.Error())
+	}
+
+	var pubKey *ecdh.PublicKey
+	if pk, ok := pubKey_unCasted.(*ecdh.PublicKey); ok {
+		pubKey = pk
+	}
+
+	// Do ECDH
+	ss, err := privKey.ECDH(pubKey)
+
+	// Kid is derived from the private key's Kid
+
+	symmetricJWK := &JWK{
+		Kty:     "EC",
+		Key_ops: []string{"encrypt", "decrypt"},
+		Kid:     "shared_" + privateKey.Kid[4:],
+		Crv:     privateKey.Crv,
+		X:       base64.URLEncoding.EncodeToString(ss),
+	}
+
+	// convert the result to a JWK.
+	return symmetricJWK, err
```

Comment:
We can have `err` as `nil`

Reply:
Agreed here also, but if we do the error checks before also, then till this step, if `err==nil`, then it will come to this point and we know that the `err` that we are returning is nil, so it's the same, but returning `nil` at the end of function instead of an err is a good practice and clean code.